psydac/api/tests/test_api_2d_compatible_spaces.py

```diff
@@ -0,0 +1,175 @@
+# -*- coding: UTF-8 -*-
+
+from sympy import pi, cos, sin, Tuple, Matrix
+
+from sympde.core import Constant
+from sympde.calculus import grad, dot, inner, cross, rot, curl, div
+from sympde.calculus import laplace, hessian
+from sympde.topology import (dx, dy, dz)
+from sympde.topology import FunctionSpace, VectorFunctionSpace
+from sympde.topology import ScalarField, VectorField
+from sympde.topology import ProductSpace
+from sympde.topology import ScalarTestFunction
+from sympde.topology import VectorTestFunction
+from sympde.topology import Boundary, NormalVector, TangentVector
+from sympde.topology import Domain, Line, Square, Cube
+from sympde.topology import Trace, trace_0, trace_1
+from sympde.topology import Union
+from sympde.expr import BilinearForm, LinearForm
+from sympde.expr import Norm, TerminalExpr
+from sympde.expr import find, EssentialBC
+
+
+from psydac.fem.basic   import FemField
+from psydac.fem.vector   import VectorFemField
+from psydac.api.discretization import discretize
+
+from numpy import linspace, zeros, allclose
+import numpy as np
+from mpi4py import MPI
+import pytest
+#np.set_printoptions(linewidth=1000, precision=4)
+from scipy.sparse.linalg import cg, gmres
+from scipy import linalg
+
+import matplotlib.pyplot as plt
+from mpl_toolkits.mplot3d import Axes3D
+from matplotlib import animation
+#==============================================================================
+
+def run_system_1_2d_dir(f0, ncells, degree):
+    # ... abstract model
+    domain = Square()
+
+    V1 = VectorFunctionSpace('V1', domain, kind='Hdiv')
+    V2 = FunctionSpace('V2', domain, kind='L2')
+    X  = ProductSpace(V1, V2)
+
+    x,y = domain.coordinates
+
+    p,q = [VectorTestFunction(V1, name=i) for i in ['p', 'q']]
+    u,v = [ScalarTestFunction(V2, name=i) for i in ['u', 'v']]
+
+    a  = BilinearForm(((p,u),(q,v)),dot(p,q) + div(q)*u + div(p)*v )
+    l  = LinearForm((q,v), f0*v)
+
+    equation = find([p,u], forall=[q,v], lhs=a((p,u),(q,v)), rhs=l(q,v))
+ 
+    # ... create the computational domain from a topological domain
+    domain_h = discretize(domain, ncells=ncells)
+    # ...
+    # ... discrete spaces
+    V1h = discretize(V1, domain_h, degree=degree)
+    V2h = discretize(V2, domain_h, degree=degree)
+    Xh  = discretize(X , domain_h, degree=degree)
+
+    # ... dsicretize the equation using Dirichlet bc
+    ah = discretize(equation, domain_h, [Xh, Xh], symbolic_space=[X, X])
+    # ...
+    ah.assemble()
+    M   = ah.linear_system.lhs.toarray()
+    rhs = ah.linear_system.rhs.toarray()
+
+    M_inv = linalg.inv(M)
+    sol = M_inv.dot(rhs)
```

Comment:
Inverting the matrix M is a slow method of solution. Instead we could use
```
sol = linalg.solve(M, rhs)
```

Even better, we can use Scipy's sparse matrix format:
```
from scipy.sparse.linalg import spsolve

M   = ah.linear_system.lhs.tosparse()
rhs = ah.linear_system.rhs.toarray()
sol = spsolve(M, rhs)
```

The same comments applies to the 2D unit tests.